File: backend/app/services/scoring.py
"""MVP scoring model: Dirichlet/Laplace-smoothed frequency over historical
special numbers (position=7 draws only).

p_i = (count_i + alpha) / (sample_count + 49 * alpha)

Deliberately does not implement "gap since last seen" or "last-period
demotion" heuristics — see spec 4.5. Those require versioned, out-of-sample
evidence before they can be enabled.
"""
from dataclasses import dataclass

BALL_RANGE = range(1, 50)  # 1-49 inclusive
ALPHA = 1.0
MODEL_NAME = "dirichlet_frequency"
MODEL_VERSION = "1.0.0"
INSUFFICIENT_HISTORY_THRESHOLD = 100


@dataclass(frozen=True)
class BallScore:
    ball: int
    probability: float
    count: int


def score_special_numbers(history: list[int], alpha: float = ALPHA) -> list[BallScore]:
    """history: special numbers (position=7 balls) in chronological order.

    Returns all 49 balls ranked by probability desc, count desc, ball asc.
    Probabilities always sum to 1 within 1e-12.
    """
    counts = {ball: 0 for ball in BALL_RANGE}
    for ball in history:
        counts[ball] += 1

    sample_count = len(history)
    denominator = sample_count + len(BALL_RANGE) * alpha

    scores = [
        BallScore(ball=ball, probability=(counts[ball] + alpha) / denominator, count=counts[ball])
        for ball in BALL_RANGE
    ]
    scores.sort(key=lambda s: (-s.probability, -s.count, s.ball))
    return scores


def gap_since_last_seen(history: list[int], ball: int) -> int | None:
    """Number of draws since `ball` last appeared as the special number.

    0 means it appeared in the most recent draw. None means it has never
    appeared in the given history.
    """
    for gap, seen in enumerate(reversed(history)):
        if seen == ball:
            return gap
    return None


def is_insufficient_history(sample_count: int) -> bool:
    return sample_count < INSUFFICIENT_HISTORY_THRESHOLD

File: atlas-ui/src/lib/scoring.ts
// Dirichlet/Laplace-smoothed frequency scoring over historical special
// numbers (position=7 draws only). Ported from backend/app/services/scoring.py
// so recommendations can be computed client-side from the static snapshot.
//
// p_i = (count_i + alpha) / (sample_count + 49 * alpha)

export const BALL_COUNT = 49;
export const ALPHA = 1.0;
export const MODEL_NAME = 'dirichlet_frequency';
export const MODEL_VERSION = '1.0.0';
export const INSUFFICIENT_HISTORY_THRESHOLD = 100;

export interface BallScore {
  ball: number;
  probability: number;
  count: number;
}

/** history: special numbers (position=7 balls) in chronological order (oldest first). */
export function scoreSpecialNumbers(history: number[], alpha: number = ALPHA): BallScore[] {
  const counts = new Map<number, number>();
  for (let ball = 1; ball <= BALL_COUNT; ball++) counts.set(ball, 0);
  for (const ball of history) counts.set(ball, (counts.get(ball) ?? 0) + 1);

  const sampleCount = history.length;
  const denominator = sampleCount + BALL_COUNT * alpha;

  const scores: BallScore[] = [];
  for (let ball = 1; ball <= BALL_COUNT; ball++) {
    const count = counts.get(ball) ?? 0;
    scores.push({ ball, probability: (count + alpha) / denominator, count });
  }
  scores.sort((a, b) => b.probability - a.probability || b.count - a.count || a.ball - b.ball);
  return scores;
}

/** 0 means it appeared in the most recent draw. null means never seen. */
export function gapSinceLastSeen(history: number[], ball: number): number | null {
  for (let gap = 0; gap < history.length; gap++) {
    if (history[history.length - 1 - gap] === ball) return gap;
  }
  return null;
}

export function isInsufficientHistory(sampleCount: number): boolean {
  return sampleCount < INSUFFICIENT_HISTORY_THRESHOLD;
}
